### PR TITLE
Fix dh_rfc5114 option in genpkey.

### PR DESCRIPTION
**Pull request by slontis**
Updated documentation for app to indicate what options are available for
DH and DHX keys. Added CHANGES entry to indicate the breaking change.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/ctrl_params_translate.c [hunk 1]
```diff
@@ -1008,6 +1008,38 @@ static int fix_dh_nid(enum state state,
...
+    int ret;
+
+    if ((ret = default_check(state, translation, ctx)) <= 0)
+        return ret;
+
+    /* This is currently only settable */
+    if (ctx->action_type != SET)
+        return 0;
+
+    if (state == PRE_CTRL_STR_TO_PARAMS) {
```
Comment (slontis) [inline crypto/evp/ctrl_params_translate.c:1025]: @levitte - Have I done this right?? It seems there are no other case of this PRE_CTRL_STR_TO_PARAMS being used like this, so either I am doing something wrong or there is possible wrongness with other entries.

File: CHANGES.md [hunk 2]
```diff
@@ -23,6 +23,21 @@ OpenSSL 3.0
 
 ### Changes between 1.1.1 and 3.0 [xx XXX xxxx]
 
+ * DH and DHX are seperate key types that are created within a provider. These
```
Comment (t8m) [inline CHANGES.md:26]: Typo: separate

File: CHANGES.md [hunk 3]
```diff
@@ -23,6 +23,21 @@ OpenSSL 3.0
...
+   application can be used to create DH domain parameters using the options
+   -genparam and -algorithm [DH or DHX]. The application then applies other
+   options using the created key. Some options are only valid for DH keys, and
+   others may be valid for DHX keys. In previous versions of OpenSSL it would
+   ignore these inconsistencies so that specifying -algorithm DH could actually
+   generate a DHX key for some options. Some of these inconsistencies may now
+   result in errors. See the openssl-genpkey(1) for further details on what
+   options should be used for the 2 key types. In general it is recommended to
+   not generate DH parameters at all, but instead to just pass a named group
+   such as ffdhe2048 to the key generation using the option -group.
```
Comment (t8m) [inline CHANGES.md:37]: I am not sure such a long changes entry is needed. I suppose just one or two summary sentences and link to manpage or the migration guide should be sufficient.

File: crypto/dh/dh_pmeth.c [hunk 4]
```diff
@@ -331,36 +330,16 @@ static int pkey_dh_paramgen(EVP_PKEY_CTX *ctx,
     /*
      * Look for a safe prime group for key establishment. Which uses
      * either RFC_3526 (modp_XXXX) or RFC_7919 (ffdheXXXX).
+     * RFC_5114 is also handles here for param_nid = (1..3)
```
Comment (t8m) [inline crypto/dh/dh_pmeth.c:333]: Typo: handled

File: doc/man1/openssl-genpkey.pod.in [hunk 5]
```diff
@@ -313,25 +413,49 @@ Generate DSA key from parameters:
...
+     -pkeyopt pbits:1024 -pkeyopt qbits:160 -pkeyopt digest:SHA1 \
+     -pkeyopt gindex:1 -pkeyopt dh_paramgen_type:1
+
+Output 2048 bit DH parameters:
 
  openssl genpkey -genparam -algorithm DH -out dhp.pem \
      -pkeyopt dh_paramgen_prime_len:2048
 
-Generate 2048 bit X9.42 DH parameters:
+Output 2048 bit X9.42 DH parameters using a generator:
```
Comment (t8m) [inline doc/man1/openssl-genpkey.pod.in:453]: These are not X9.42 DH but plain DH.

**Review by t8m — CHANGES_REQUESTED** [review]
Just minor nits, otherwise I like this very much.

File: doc/man1/openssl-genpkey.pod.in (hunk 5 above)
Comment (FdaSilvaYY) [reply doc/man1/openssl-genpkey.pod.in:453]: nit : 2048 bit**S** ?

File: CHANGES.md (hunk 2 above)
Comment (slontis) [reply CHANGES.md:26]: yeah I struggle with that word :)

File: doc/man1/openssl-genpkey.pod.in (hunk 5 above)
Comment (slontis) [reply doc/man1/openssl-genpkey.pod.in:453]: Don't think it needs the s

**Comment by slontis** [comment]
Quite a few changes in the last 3 commits - had to rebase to update the commit message..

**Comment by slontis** [comment]
This errors in the setparams if bad parameters are passed in. If we dont do this then potentially it would do something different from what was expected.

**Comment by t8m** [comment]
The CI failures are relevatnt.

**Comment by t8m** [comment]
(Except for the external-tests, that is a different thing)

File: crypto/ffc/ffc_dh.c [hunk 6]
```diff
@@ -113,9 +113,7 @@ const DH_NAMED_GROUP *ossl_ffc_numbers_to_dh_named_group(const BIGNUM *p,
         if (BN_cmp(p, dh_named_groups[i].p) == 0
             && BN_cmp(g, dh_named_groups[i].g) == 0
             /* Verify q is correct if it exists */
-            && ((q != NULL && BN_cmp(q, dh_named_groups[i].q) == 0)
-                /* Do not match RFC 5114 groups without q */
-                || (q == NULL && dh_named_groups[i].uid > 3)))
```
Comment (slontis) [inline crypto/ffc/ffc_dh.c:118]: NOTE - this line was a hack to fix a decoder test issue - the root cause was fixed so this is no longer needed. This allows RFC5114 to be able to be loaded correctly as a named group. (The original problem was that the DH asn1 callback was not setting the nid - but it was for DHX asn1)

File: crypto/evp/ctrl_params_translate.c (hunk 1 above)
Comment (t8m) [reply crypto/evp/ctrl_params_translate.c:1017]: IMO we do not have any other CTRLs where you convert an integer value into a named parameter.

File: doc/man7/EVP_PKEY-DH.pod [hunk 7]
```diff
@@ -14,25 +15,30 @@ applications that cannot be upgraded to use the approved safe-prime groups.
...
+For B<DHX> that is not a named group the FIPS186-4 standard specifies that the
 values used for FFC parameter generation are also required for parameter
 validation. This means that optional FFC domain parameter values for
-I<seed>, I<pcounter> and I<gindex> may need to be stored for validation purposes.
-For B<DH> the I<seed> and I<pcounter> can be stored in ASN1 data
-(but the I<gindex> is not).
+I<seed>, I<pcounter> and I<gindex> or I<hindex> may need to be stored for
+validation purposes.
+For B<DHX> the I<seed> and I<pcounter> can be stored in ASN1 data
+(but the I<gindex> or I<hindex> is not).
```
Comment (t8m) [inline doc/man7/EVP_PKEY-DH.pod:28]: s/is not/can not/

File: doc/man7/EVP_PKEY-DH.pod [hunk 8]
```diff
@@ -14,25 +15,30 @@ applications that cannot be upgraded to use the approved safe-prime groups.
...
+(see L<EVP_PKEY-FFC(7)/FFC parameters>) the B<DHX> and B<DH> keytype
+implementation support the following:
 
 =over 4
 
 =item "group" (B<OSSL_PKEY_PARAM_GROUP_NAME>) <UTF8 string>
 
-Set or gets a string that associates a B<DH> named safe prime group with known
-values for I<p>, I<q> and I<g>.
+Set or gets a string that associates a B<DH> or B<DHX> named safe prime group
```
Comment (t8m) [inline doc/man7/EVP_PKEY-DH.pod:40]: Sets or gets

File: doc/man7/EVP_PKEY-DH.pod [hunk 9]
```diff
@@ -41,31 +47,45 @@ The following values can be used by the OpenSSL's default and FIPS providers:
...
+=over 4
+
+=item "safeprime-generator" (B<OSSL_PKEY_PARAM_DH_GENERATOR>) <integer>
+
+Used for DH generation of safe primes using the old safe prime generator code.
+The default value is 2.
+It is recommended to use a named safe prime group instead, if domain parameter
+validation is required. 
+
+These are not 'named' safe prime groups so setting this value for the OpenSSL FIPS
```
Comment (t8m) [inline doc/man7/EVP_PKEY-DH.pod:79]: Maybe: `Randomly generated safe prime groups are not allowed by FIPS so setting ...`

File: providers/implementations/keymgmt/dh_kmgmt.c [hunk 10]
```diff
@@ -519,11 +518,59 @@ static int dh_gen_set_params(void *genctx, const OSSL_PARAM params[])
...
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_DH_PRIV_LEN);
+    if (p != NULL && !OSSL_PARAM_get_int(p, &gctx->priv_len))
+        return 0;
+    return 1;
+}
+
+static const OSSL_PARAM *dh_gen_settable_params(ossl_unused void *genctx,
+                                                ossl_unused void *provctx)
+{
+    static OSSL_PARAM dh_gen_settable[] = {
```
Comment (t8m) [inline providers/implementations/keymgmt/dh_kmgmt.c:534]: Should it be static const?

File: providers/implementations/keymgmt/dh_kmgmt.c [hunk 11]
```diff
@@ -519,11 +518,59 @@ static int dh_gen_set_params(void *genctx, const OSSL_PARAM params[])
...
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_DH_GENERATOR, NULL),
+        OSSL_PARAM_END
+    };
+    return dh_gen_settable;
+}
+
+static const OSSL_PARAM *dhx_gen_settable_params(ossl_unused void *genctx,
+                                                 ossl_unused void *provctx)
+{
+    static OSSL_PARAM dhx_gen_settable[] = {
```
Comment (t8m) [inline providers/implementations/keymgmt/dh_kmgmt.c:548]: Same here

File: doc/man7/EVP_PKEY-DH.pod [hunk 12]
```diff
@@ -14,25 +15,30 @@ applications that cannot be upgraded to use the approved safe-prime groups.
...
+(but the I<gindex> or I<hindex> is not).
 
-=head2 DH parameters
+=head2 DH and DHX domain parameters
 
 In addition to the common FCC parameters that all FFC keytypes should support
-(see L<EVP_PKEY-FFC(7)/FFC parameters>) the B<DH> keytype
-implementation supports the following:
+(see L<EVP_PKEY-FFC(7)/FFC parameters>) the B<DHX> and B<DH> keytype
+implementation support the following:
```
Comment (t8m) [inline doc/man7/EVP_PKEY-DH.pod:34]: implementations

File: doc/man1/openssl-genpkey.pod.in [hunk 13]
```diff
@@ -243,36 +251,128 @@ generate valid primes.
...
-1, 2 or 3 corresponding to RFC5114 DH parameters consisting of
-1024 bit group with 160 bit subgroup, 2048 bit group with 224 bit subgroup
-and 2048 bit group with 256 bit subgroup as mentioned in RFC5114 sections
-2.1, 2.2 and 2.3 respectively. If present this overrides all other DH parameter
-options.
+The digest to use during parameter generation. Must be one of B<sha1>, B<sha224>
+or B<sha256>. If set, then the number of bits in B<qbits> will match the output
+size of the specified digest and the B<qbits> parameter will be
+ignored. If not set, then a digest will be used that gives an output matching
+the number of bits in B<q>, i.e. B<sha1> if q length is 160, B<sha224> if it 224
```
Comment (t8m) [inline doc/man1/openssl-genpkey.pod.in:351]: if it is 224

**Review by t8m — CHANGES_REQUESTED** [review]
A few more minor nits otherwise LGTM

**Comment by slontis** [comment]
NITS addressed

**Review by t8m — APPROVED** [review]
Great work!

**Comment by t8m** [comment]
Merged to master. I've added a new title line of the commit message as the original one did not really cover all the change. Thank you for the contribution, Shane!

**Pull request closed without merging**